budget/app.py

```diff
@@ -106,8 +137,93 @@ def compare_budgets_and_users(account_id, users):
 
 def create_budget(synapse_id, team):
   '''Creates an AWS budget for a synapse user id'''
-  # TODO implement
-  log.info('Placeholder -- this will make a budget')
+  log.debug(f'Creating budget for synapse user {synapse_id}, member of team {team}')
+
+  team_budget_rules = TEAM_BUDGET_RULES.get('teams').get(team)
+  if not team_budget_rules:
+    raise ValueError(f'No budget rules available for team {team}')
+  budget_amount = team_budget_rules['amount']
+  budget_period = team_budget_rules['period']
+  budget = {
+    'BudgetName': f'{BUDGET_NAME_PREFIX}{synapse_id}',
+    'BudgetLimit': {
+      'Amount': budget_amount,
+      'Unit': 'USD'
+    },
+    'CostFilters': {
+      'TagKeyValue': [
+        (
+          'aws:servicecatalog:provisioningPrincipalArn$arn:aws:sts::'
+          f'{_get_account_id()}:assumed-role/ServiceCatalogEndusers/{synapse_id}'
+        )
+      ]
+    },
+    'CostTypes': {
+      'IncludeRefund': False,
+      'IncludeCredit': False
+    },
+    'TimeUnit': budget_period,
+    'BudgetType': 'COST'
+  }
+  budgets_client = get_client('budgets')
+  return budgets_client.create_budget(
+    AccountId=_get_account_id(),
+    Budget=budget
+    )
+
+
+def _create_budget_notification(synapse_id, threshold, admin_emails=None):
+  '''Creates a notification rule when the budget crosses a particular threshold.
+
+  May have multiple recipients. User notifications are sent through SNS
+  in order to send email to Synapse addresses, which cannot receive email
+  unless sent through the Synapse system, but emails to administrators can
+  be sent directly.
+  '''
+  budgets_client = get_client('budgets')
+  budget_name = f'{BUDGET_NAME_PREFIX}{synapse_id}'
+  subscribers = []
+  if admin_emails:
+    subscribers = [{
+      'SubscriptionType': 'EMAIL',
+      'Address': address
+    } for address in admin_emails]
+  subscribers.append({
+    'SubscriptionType': 'SNS',
+    'Address': _get_notification_topic_arn()
+    })
+
+  return budgets_client.create_notification(
+      AccountId=_get_account_id(),
+      BudgetName=budget_name,
+      Notification={
+          'NotificationType': 'ACTUAL',
+          'ComparisonOperator': 'GREATER_THAN',
+          'Threshold': threshold,
+          'ThresholdType': 'PERCENTAGE',
+          'NotificationState': 'ALARM'
+      },
+      Subscribers=subscribers
+  )
+
+
+def create_budget_notifications(synapse_id, team):
+  '''Creates notifications for a particular budget'''
+  budget_name = f'{BUDGET_NAME_PREFIX}{synapse_id}'
+  thresholds = TEAM_BUDGET_RULES['thresholds']
+
+  for threshold in thresholds['notify_user_only']:
+    _create_budget_notification(
+      synapse_id,
+      threshold
+    )
+
+  for threshold in thresholds['notify_admins_too']:
+    _create_budget_notification(
+      synapse_id,
+      threshold,
+      admin_emails=TEAM_BUDGET_RULES['teams'][team]['community_manager_emails']
```

Comment:
Will we also need to tell the community manager which team is causing the alarm? maybe team membership is mutually exclusive for all team members, but a community manager may have multiple communities to watch over...

Reply:
The alarm will be for a particular user. What's in the email will have the budget name, which includes the synapse id, that it passed a certain threshold, and what the total budget amount is for the user. It might be a nice improvement to include the team id in the budget name as well. I'll consider whether we have time to include that for MVP, but there are competing priorities.

Reply:
I believe we assume one team per user, so we can always look up the team from userid when needed.

Reply:
Correct, this lambda assumes that external users will be in only one of the teams.